wicker/l5ml_datastore/v2/index_ops.py

```diff
@@ -0,0 +1,253 @@
+    """
+    if left.primary_keys != right.primary_keys:
+        raise WickerSchemaException(
+            "The schemas have different primary keys: Got " f"{left.primary_keys} and  {right.primary_keys}"
+        )
+    # Make deep copies of the schemas so that we can change the "required" status.
+    left = copy.deepcopy(left)
+    right = copy.deepcopy(right)
+
+    new_fields = left._columns
+    if not left_required:
+        for name, field in new_fields.items():
+            field.required = name in left.primary_keys
+    for name, field in right._columns.items():
+        if name in new_fields:
+            if not new_fields[name].is_compatible(field):
+                raise WickerSchemaException(
+                    f"Can not merge schemas. Field '{name}' seen with two different types. "
+                    f"{type(new_fields[name])}  {type(field)}"
+                )
+            if right_required and field.required:
+                new_fields[name].required = True
+        else:
+            field.required = field.required and right_required
+            new_fields[name] = field
+    return DatasetSchema(fields=list(new_fields.values()), primary_keys=left.primary_keys)
+
+
+def join_datasets(
+    left: AbstractDataset[Example], right: AbstractDataset[Example], how: str = "outer"
+) -> AbstractDataset[Example]:
+    """Join two datasets (that have the same primary keys) on their primary keys.
+    :param left: First dataset to merge.
+    :param right: Second dataset to merge.
+    :param how: Type of merge to apply. The possible values follow the options for pandas.Dataframe.merge:
+      "left": use only keys from left frame, similar to a SQL left outer join
+      "right": use only keys from right frame, similar to a SQL right outer join
+      "outer": use union of keys from both frames, similar to a SQL full outer join
+      "inner": use intersection of keys from both frames, similar to a SQL inner join
+    :return: A new dataset resulting from the merge of "left" and "right".
+    """
+    # Note, this function does a very simple merge based on equality of keys.
+    # Maybe we should have a function for resolving approximate timestamps.
+    # We could truncate the precision of timestamps, but that's not a very good solution.
+    # Better to use the left side's timestamp as reference and merge right side rows that are within an
+    # error threshold.
+    if not left.is_compatible(right):
+        raise WickerDatastoreException(
+            "Dataset a and b are not compatible. Datasets must be of the same type "
+            "and have the same backing infrastructure."
+        )
+    left_required = False
+    right_required = False
+    if how == "inner":
+        left_required = True
+        right_required = True
+    elif how == "left":
+        left_required = True
+    elif how == "right":
+        right_required = True
+    elif how == "outer":
+        pass
```

Comment:
I wonder if it would make more sense to have a function for each that is a basic wrapper on this function, do you think that would be more clear?